### PR TITLE
Status changed by User on the task will be logged

### PR DESCRIPTION
**Pull request by ivinayakg**
# Parent RFC - #460 

## How will this work?
- Will create a new log and save to the database about the event

## What has been done?
- called the `addlog` API to create a log for the event
- Response after a successful call
image

## Edge Cases Covered
- If the user is trying to change the status to `MERGED`, it will be forbidden.
- If `status` is already `COMPLETED` and the user is trying to change only the `percentCompleted` to less than 100, it will give a bad request.
- If `status` is not `COMPLETED` and the user is trying to change the `percentCompleted` to 100, it will give a bad request.

### CODE REVIEW

File: controllers/tasks.js [hunk 1]
```diff
@@ -142,23 +143,44 @@ const updateTaskStatus = async (req, res, next) => {
...
+        userId,
+        taskId,
+        username,
+      },
+      body: [],
+    };
+
+    if (req.body.status) {
+      taskLog.body.push(`${username} changed task status to ${req.body.status}`);
+    }
```
Comment (vinayak-trivedi) [inline controllers/tasks.js:166]: if in the frontend I want to check the value of status, then I wouldn't be able to check it right? I think we should send the status as well with this. what do you think?

File: controllers/tasks.js [hunk 2]
```diff
@@ -142,23 +143,48 @@ const updateTaskStatus = async (req, res, next) => {
...
+      },
+    };
+
+    if (req.body.status) {
+      taskLog.body.messages.push(`${username} changed task status to ${req.body.status}`);
+      taskLog.body.newStatus = req.body.status;
+    }
+    if (req.body.percentCompleted) {
+      taskLog.body.messages.push(`${username} changed task percent Completed to ${req.body.percentCompleted}`);
+      taskLog.body.newPercentCompleted = req.body.percentCompleted;
```
Comment (vinayak-trivedi) [inline controllers/tasks.js:172]: do we need to put username here as we are already sending the username in meta above

File: test/integration/tasks.test.js [hunk 3]
```diff
@@ -394,6 +402,79 @@ describe("Tasks", function () {
...
+        endsOn: 1234,
+        startedOn: 4567,
+        status: "REVIEW",
+        percentCompleted: 10,
+        participants: [],
+        assignee: appOwner.username,
+        completionAward: { [DINERO]: 3, [NEELAM]: 300 },
+        lossRate: { [DINERO]: 1 },
+        isNoteworthy: true,
+      };
```
Comment (RitikJaiswal75) [inline test/integration/tasks.test.js:442]: Please store this data in a fixture and pass it here rather than retyping it everywhere

dynamic data can be passed as a parameter

File: controllers/tasks.js [hunk 4]
```diff
@@ -142,23 +143,59 @@ const updateTaskStatus = async (req, res, next) => {
...
     const taskId = req.params.id;
     const { dev } = req.query;
-    const { id: userId } = req.userData;
+    const { id: userId, username } = req.userData;
     const task = await tasks.fetchSelfTask(taskId, userId);
 
     if (task.taskNotFound) return res.boom.notFound("Task doesn't exist");
     if (task.notAssignedToYou) return res.boom.forbidden("This task is not assigned to you");
-    if (task.taskData.status === "VERIFIED")
+    if (task.taskData.status === "VERIFIED" || req.body.status === "MERGED")
```
Comment (RitikJaiswal75) [inline controllers/tasks.js:151]: image
Are all these states being checked?
like for needs review Percentage should be > 80

File: test/integration/tasks.test.js [hunk 5]
```diff
@@ -313,6 +313,19 @@ describe("Tasks", function () {
...
+      type: "feature",
+      endsOn: 1234,
+      startedOn: 4567,
+      status: "VERIFIED",
+      percentCompleted: 10,
+      participants: [],
+      completionAward: { [DINERO]: 3, [NEELAM]: 300 },
+      lossRate: { [DINERO]: 1 },
+      isNoteworthy: true,
+    };
```
Comment (RitikJaiswal75) [inline test/integration/tasks.test.js:327]: This is being re-used a lot create a function in fixtures and call it with valid statuses

File: test/integration/tasks.test.js [hunk 6]
```diff
@@ -371,13 +392,47 @@ describe("Tasks", function () {
...
+      const res = await chai
+        .request(app)
+        .patch(`/tasks/self/${taskId}`)
+        .set("cookie", `${cookieName}=${jwt}`)
+        .send(taskStatusData);
+
+      expect(res).to.have.status(403);
+      expect(res.body.message).to.be.equal("Status cannot be updated. Please contact admin.");
+    });
+    it("Should give 403 if new status is 'MERGED' ", async function () {
```
Comment (RitikJaiswal75) [inline test/integration/tasks.test.js:405]: An HTTP 403 response code means that a client is forbidden from accessing a valid URL

Does status 403 hold good here? if yes can you explain

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403

File: test/integration/tasks.test.js [hunk 7]
```diff
@@ -371,13 +392,47 @@ describe("Tasks", function () {
...
+        .request(app)
+        .patch(`/tasks/self/${taskId}`)
+        .set("cookie", `${cookieName}=${jwt}`)
+        .send({ ...taskStatusData, status: "COMPLETED" });
+
+      expect(res).to.have.status(403);
+      expect(res.body.message).to.be.equal("Status cannot be updated. Task is not completed yet");
+    });
+
+    it("Should give 403 if status is COMPLETED and newpercent is less than 100", async function () {
```
Comment (RitikJaiswal75) [inline test/integration/tasks.test.js:428]: how is it different from above test, please make a proper sentence

File: controllers/tasks.js [hunk 8]
```diff
@@ -142,23 +143,59 @@ const updateTaskStatus = async (req, res, next) => {
...
+        new: {},
+      },
+    };
+
+    if (req.body.status) {
+      taskLog.body.new.status = req.body.status;
+    }
+    if (req.body.percentCompleted) {
+      taskLog.body.new.percentCompleted = req.body.percentCompleted;
+    }
```
Comment (yesyash) [inline controllers/tasks.js:184]: In case of `req.body.status` & `req.body.percentCompleted` not being present then `taskLog.body.new.status` & `taskLog.body.new.percentCompleted` will be undefined, is this okay ?

Comment (ivinayakg) [reply controllers/tasks.js:184]: There are if statements for that reason only.

Comment (yesyash) [reply controllers/tasks.js:184]: What I meant was are `taskLog.boy.new.status` & `taskLog.body.new.percentCompleted` optional fields ? Is it fine okay if we do not have them ?

File: controllers/tasks.js [hunk 9]
```diff
@@ -142,23 +143,59 @@ const updateTaskStatus = async (req, res, next) => {
...
 
     if (task.taskNotFound) return res.boom.notFound("Task doesn't exist");
     if (task.notAssignedToYou) return res.boom.forbidden("This task is not assigned to you");
-    if (task.taskData.status === "VERIFIED")
+    if (task.taskData.status === "VERIFIED" || req.body.status === "MERGED")
       return res.boom.forbidden("Status cannot be updated. Please contact admin.");
 
-    await tasks.updateTask(req.body, taskId);
+    if (task.taskData.status === "COMPLETED" && req.body.percentCompleted < 100) {
+      if (req.body.status === "COMPLETED" || !req.body.status) {
```
Comment (pallabez) [inline controllers/tasks.js:155]: What are we checking? Can we add a comment here?

File: controllers/tasks.js [hunk 10]
```diff
@@ -142,23 +143,59 @@ const updateTaskStatus = async (req, res, next) => {
...
+      tasks.updateTask(req.body, taskId),
+      addLog(taskLog.type, taskLog.meta, taskLog.body),
+    ]);
+    taskLog.id = taskLogResult.id;
 
     if (dev) {
       if (req.body.percentCompleted === 100) {
         return next();
       }
     }
```
Comment (pallabez) [inline controllers/tasks.js:196]: What does it do? Is it a middleware?

File: controllers/tasks.js (hunk 8 above)
Comment (ivinayakg) [reply controllers/tasks.js:184]: no, it wouldn't be as the approach was to log the task event in a manner that the raw data should be self-explanatory about what was the event that happened with specificity

File: controllers/tasks.js (hunk 10 above)
Comment (ivinayakg) [reply controllers/tasks.js:196]: not part of this PR, I belive this is something to do with task leveling

File: controllers/tasks.js (hunk 9 above)
Comment (ivinayakg) [reply controllers/tasks.js:155]: im trying to cover as many edge cases as possible, so this is the code to cover one of the edge cases.
what do you think should we write docs for it or write comments there?

File: test/integration/tasks.test.js (hunk 6 above)
Comment (ivinayakg) [reply test/integration/tasks.test.js:405]: yes, it is good here as I'm trying to forbid the user to put the status as merged 
I want to give that functionality only to the super-user only

File: test/integration/tasks.test.js (hunk 7 above)
Comment (ivinayakg) [reply test/integration/tasks.test.js:428]: they are very different check again

File: controllers/tasks.js (hunk 4 above)
Comment (ivinayakg) [reply controllers/tasks.js:151]: for that i guess we should have a proper discussion with other members on the call

File: controllers/tasks.js [hunk 11]
```diff
@@ -142,23 +143,59 @@ const updateTaskStatus = async (req, res, next) => {
...
-    if (task.taskData.status === "VERIFIED")
+    if (task.taskData.status === "VERIFIED" || req.body.status === "MERGED")
       return res.boom.forbidden("Status cannot be updated. Please contact admin.");
 
-    await tasks.updateTask(req.body, taskId);
+    if (task.taskData.status === "COMPLETED" && req.body.percentCompleted < 100) {
+      if (req.body.status === "COMPLETED" || !req.body.status) {
+        return res.boom.forbidden("Task percentCompleted can't updated as status is COMPLETED");
+      }
+    }
```
Comment (ivinayakg) [inline controllers/tasks.js:158]: if the task status is `COMPLETED` and the user is trying to change the `percentCompleted` with a value of less than 100, if so then to check wether the user is also trying to change the status of the task to something else then `COMPLETED` otherwise will give an error.
### Basically the status `COMPLETED` should be there only with percentCompleted as `100`

File: controllers/tasks.js [hunk 12]
```diff
@@ -142,23 +143,59 @@ const updateTaskStatus = async (req, res, next) => {
...
+
+    if (req.body.status === "COMPLETED" && task.taskData.percentCompleted !== 100) {
+      if (req.body.percentCompleted !== 100) {
+        return res.boom.forbidden("Status cannot be updated. Task is not completed yet");
+      }
+    }
```
Comment (ivinayakg) [inline controllers/tasks.js:164]: If the user is trying to change the status only with the value of `COMPLETED` and the value of percentCompleted is not 100 and the new value (if so) is also not 100 then will give an error.

### Basically the status `COMPLETED` should be there only with percentCompleted as `100`

File: controllers/tasks.js (hunk 4 above)
Comment (Pavangbhat) [inline controllers/tasks.js:151]: Let's please use constants

**Review by Pavangbhat — CHANGES_REQUESTED** [review]
Please make use of constants where ever possible

File: controllers/tasks.js (hunk 4 above)
Comment (ivinayakg) [reply controllers/tasks.js:151]: can you please elaborate a bit, I didn't get what you were saying...

Comment (Pavangbhat) [reply controllers/tasks.js:151]: I meant we can place these strings in a seperate file ( can be referred as constants ) and use it. probably there is already a file for task status constants

Comment (ivinayakg) [reply controllers/tasks.js:151]: ok, got it make sense

**Review by vinayak-trivedi — DISMISSED** [review]
LGTM

**Comment by ivinayakg** [comment]
# DO NOT MERGE

**Review by Maheima — APPROVED** [review]
LGTM!

File: controllers/tasks.js [hunk 13]
```diff
@@ -157,23 +158,64 @@ const updateTaskStatus = async (req, res, next) => {
...
+    const taskLog = {
+      type: "task",
+      meta: {
+        userId,
+        taskId,
+        username,
+      },
+      body: {
+        subType: "update",
+        new: {},
```
Comment (prakashchoudhary07) [inline controllers/tasks.js:190]: Why are we passing empty object here?

**Review by prakashchoudhary07 — APPROVED** [review]

**Comment by prakashchoudhary07** [comment]
Maybe we can create a separate middleware that has all rules regarding the tasks?

File: test/integration/tasks.test.js [hunk 14]
```diff
@@ -453,13 +473,47 @@ describe("Tasks", function () {
...
+
+    it("Should give 400 if status is COMPLETED and newpercent is less than 100", async function () {
       const taskData = {
         title: "Test task",
         type: "feature",
         endsOn: 1234,
         startedOn: 4567,
-        status: "VERIFIED",
-        percentCompleted: 10,
+        status: "COMPLETED",
```
Comment (ankushdharkar) [inline test/integration/tasks.test.js:515]: Any constants should come from a constants file. "MERGED", "COMPLETED", etc

Comment (ankushdharkar) [reply test/integration/tasks.test.js:515]: (Commented to create the placeholder for #859)
